### PR TITLE
opencolorio: use system lcms2 on aarch64

### PR DESCRIPTION
**Pull request by Thra11**
The version of LCMS bundled with opencolorio is too old to build on
aarch64, simply because its config files date from before aarch64
was announced. However, it can use the system lcms2 if it is found.

###### Motivation for this change
OpenColorIO fails to build on AArch64.

###### Things done
* Added cmake flag to disable SSE on non x86 platforms.
* Added cmake flag on arm and aarch64 which in theory forces it to use the system LCMS (In practice, it appears to be defined but unused, but I figured it should be set to the correct value in case it gets fixed in future).
* Added pkgconfig as a native build input and lcms2 as a buildInput on aarch64, so OCIO's cmake config can find this 'system' lcms2 instead of using the bundled one.

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

cc @cillianderoiste 

---

This will probably trigger rebuilds on i686 and x86_64 too, as I moved cmake and unzip from buildInputs to nativeBuildInputs.
We could probably switch to building against system LCMS on all platforms if we wanted to simplify things by removing all the `optional` bits. However, I haven't tested that.

### CODE REVIEW

File: pkgs/development/libraries/opencolorio/default.nix [hunk 1]
```diff
@@ -11,9 +13,15 @@ stdenv.mkDerivation rec {
...
+  nativeBuildInputs = [ cmake unzip ]
+                    ++ optional (stdenv.hostPlatform.isAarch32 || stdenv.hostPlatform.isAarch64) pkgconfig;
+
+  buildInputs = optional stdenv.isDarwin boost
+              ++ optional (stdenv.hostPlatform.isAarch32 || stdenv.hostPlatform.isAarch64) lcms2;
 
-  cmakeFlags = lib.optional stdenv.isDarwin "-DOCIO_USE_BOOST_PTR=ON";
+  cmakeFlags = optional stdenv.isDarwin "-DOCIO_USE_BOOST_PTR=ON"
+              ++ optional (!stdenv.hostPlatform.isi686 && !stdenv.hostPlatform.isx86_64) "-DOCIO_USE_SSE=OFF"
+              ++ optional (stdenv.hostPlatform.isAarch32 || stdenv.hostPlatform.isAarch64) "-DUSE_EXTERNAL_LCMS=ON";
```
[the remark this answers is not in the document]
Comment (Thra11) [reply pkgs/development/libraries/opencolorio/default.nix]: Shall I change this pull request / make another one to do that then?
While we're at it, should we go non-vendor-crazy and try using system yaml and system tinyxml as well?

[the remark this answers is not in the document]
Comment (Mic92) [reply pkgs/development/libraries/opencolorio/default.nix]: No, this pull request is fine. Yaml/tinyxml can be also de-vendored.

**Comment by Thra11** [comment]
No problem de-vendoring tinyxml as well. The current version of libyamlcpp is incompatible (changes in iterator API), so we can either stick with the vendored yaml, build an older system yaml, or patch ocio to work with the more recent yaml (I haven't seen any such patches in the wild yet).

Another minor issue I noticed is that the last commit (https://github.com/NixOS/nixpkgs/commit/8ea7ad67fd19f3fc7866084c8648b6fd2d1c79e9#diff-2e251a0e479090d497b6670fee265f97) bumps the package version but doesn't actually change the sources (version is hardcoded in the fetch url). If I do change the fetch url to 1.1.0, it starts requiring git if using system lcms (haven't looked into why yet, something to do with patches) and no longer compiles.

My inclination is to, at least initially, take the lazy route:
* De-vendor lcms and tinyxml
* Drop the package version back down to match the source url

**Comment by jtojnar** [comment]
> The current version of libyamlcpp is incompatible (changes in iterator API), so we can either stick with the vendored yaml, build an older system yaml, or patch ocio to work with the more recent yaml (I haven't seen any such patches in the wild yet).

For now I would add `-DUSE_EXTERNAL_YAML=off` with a comment linking to https://github.com/imageworks/OpenColorIO/issues/517. If you are able to provide a patch, that would be awesome, but using vendored yaml would do for now until upstream is fixed.

> Another minor issue I noticed is that the last commit (https://github.com/NixOS/nixpkgs/commit/8ea7ad67fd19f3fc7866084c8648b6fd2d1c79e9#diff-2e251a0e479090d497b6670fee265f97) bumps the package version but doesn't actually change the sources (version is hardcoded in the fetch url). If I do change the fetch url to 1.1.0, it starts requiring git if using system lcms (haven't looked into why yet, something to do with patches) and no longer compiles.

Why not just add git for now?

**Comment by Thra11** [comment]
> Why not just add git for now?

There are a couple of other issues in 1.1.0 which cause the build to fail:
* `-Werror` is specified in `CMakeLists.txt`, but the compiler emits several warnings (assume upstream use an older compiler)
* The installPhase tries to `rm /nix/store/aq6njh14lqfwf7034y39xwf8ypglk9yr-opencolorio-1.1.0/lib/*.a`, but no files match the glob.

The first can be fixed by removing the offending `Werror`s after unpacking (I tried various permutations of `-DCMAKE_CXX_FLAGS=-Wno-error=unused-function -Wno-error-deprecated-declarations`, but I can't get it to pass the two space-separated flags through without dropping the second one)
I haven't worked out how to fix the second issue yet.

**Comment by Thra11** [comment]
Hang on! The `rm` command is coming from `postInstall`, so I can just remove it from there.

**Comment by jtojnar** [comment]
Try `CXXFLAGS` attribute in `mkDerivation`.

**Comment by Thra11** [comment]
CXXFLAGS is cleaner, thanks.

**Comment by Thra11** [comment]
@GrahamcOfBorg build opencolorio

File: pkgs/development/libraries/opencolorio/default.nix [hunk 2]
```diff
@@ -1,22 +1,33 @@
-{ stdenv, lib, fetchurl, cmake, unzip, boost }:
+{ stdenv, lib, fetchurl, cmake, unzip, boost, pkgconfig, lcms2, tinyxml, git }:
+
+with lib;
 
 stdenv.mkDerivation rec {
   name = "opencolorio-${version}";
   version = "1.1.0";
 
   src = fetchurl {
```
Comment (jtojnar) [inline pkgs/development/libraries/opencolorio/default.nix]: Try using `fetchFromGitHub` it is preferred over `fetchurl` and will probably allow you to get rid of `unzip`.

Comment (Thra11) [reply pkgs/development/libraries/opencolorio/default.nix]: Done. Thanks.

File: pkgs/development/libraries/opencolorio/default.nix [hunk 3]
```diff
@@ -1,22 +1,35 @@
...
 
-  buildInputs = [ cmake unzip ] ++ lib.optional stdenv.isDarwin boost;
+  nativeBuildInputs = [ cmake pkgconfig git ];
+
+  buildInputs = [ lcms2 tinyxml ] ++ optional stdenv.isDarwin boost;
+
+  CXXFLAGS = "-Wno-error=unused-function -Wno-error=deprecated-declarations";
 
-  cmakeFlags = lib.optional stdenv.isDarwin "-DOCIO_USE_BOOST_PTR=ON";
+  # External libyamlcpp 0.6.* not compatible: https://github.com/imageworks/OpenColorIO/issues/517
```
Comment (jtojnar) [inline pkgs/development/libraries/opencolorio/default.nix]: nit: Could you move it closer to the relevant flag?

File: pkgs/development/libraries/opencolorio/default.nix [hunk 4]
```diff
@@ -1,22 +1,35 @@
...
+    owner = "imageworks";
+    repo = "OpenColorIO";
+    rev = "v${version}";
+    sha256 = "0cjsyn681xsg89lirlll5pqlsqg2vnk1278iiicmzjy2a2v8x7zq";
   };
 
   outputs = [ "bin" "out" "dev" ];
 
-  buildInputs = [ cmake unzip ] ++ lib.optional stdenv.isDarwin boost;
+  nativeBuildInputs = [ cmake pkgconfig git ];
```
Comment (jtojnar) [inline pkgs/development/libraries/opencolorio/default.nix]: Could you try to get rid of git? Or at least add a todo comment.

Comment (Thra11) [reply pkgs/development/libraries/opencolorio/default.nix]: If I try to remove git, I get
```
CMake Error at /nix/store/48jl65nlh1msp8yi4apbj7czp1dr978k-cmake-3.11.2/share/cmake-3.11/Modules/FindPackageHandleStandardArgs.cmake:137 (message):
  Could NOT find Git (missing: GIT_EXECUTABLE)
Call Stack (most recent call first):
  /nix/store/48jl65nlh1msp8yi4apbj7czp1dr978k-cmake-3.11.2/share/cmake-3.11/Modules/FindPackageHandleStandardArgs.cmake:378 (_FPHSA_FAILURE_MESSAGE)
  /nix/store/48jl65nlh1msp8yi4apbj7czp1dr978k-cmake-3.11.2/share/cmake-3.11/Modules/FindGit.cmake:83 (find_package_handle_standard_args)
  CMakeLists.txt:339 (find_package)
```
It looks like git is being used as a glorified patch command:
```
find_package(Git REQUIRED) ## in order to apply patch (for crossplateform 
compatibility)
        ExternalProject_Add(tinyxml
            URL             ${TINYXML_ZIPFILE}
            SOURCE_DIR      ${TINYXML_SOURCE_DIR}/tinyxml
            PATCH_COMMAND   ${GIT_EXECUTABLE} apply --ignore-whitespace ${TINYXML_
PATCHFILE}
```
I've added a todo.

File: pkgs/development/libraries/opencolorio/default.nix (hunk 3 above)
Comment (Thra11) [reply pkgs/development/libraries/opencolorio/default.nix]: Done

**Comment by jtojnar** [comment]
@GrahamcOfBorg build opencolorio

File: pkgs/development/libraries/opencolorio/default.nix [hunk 5]
```diff
@@ -1,22 +1,36 @@
...
 
   outputs = [ "bin" "out" "dev" ];
 
-  buildInputs = [ cmake unzip ] ++ lib.optional stdenv.isDarwin boost;
+  # TODO: Investigate whether git can be dropped: It's only used to apply patches
+  nativeBuildInputs = [ cmake pkgconfig git ];
+
+  buildInputs = [ lcms2 tinyxml ] ++ optional stdenv.isDarwin boost;
+
+  CXXFLAGS = "-Wno-error=unused-function -Wno-error=deprecated-declarations";
```
Comment (Mic92) [inline pkgs/development/libraries/opencolorio/default.nix]: Can disable `-Werror` all together? This is a maintenance nightmare when upgrading gcc.

Comment (jtojnar) [reply pkgs/development/libraries/opencolorio/default.nix]: And could you also report the `-Werror` issue to uptream and add a comment with a link here.

Comment (Mic92) [reply pkgs/development/libraries/opencolorio/default.nix]: They can set it explicit in CFLAGS when running their CI build: https://github.com/imageworks/OpenColorIO/blob/master/.travis.yml
Apart from that they should default `-Wall`.

Comment (Mic92) [reply pkgs/development/libraries/opencolorio/default.nix]: http://blog.schmorp.de/2016-02-27-tidbits-for-the-love-of-god-dont-use-werror.html

Comment (Thra11) [reply pkgs/development/libraries/opencolorio/default.nix]: Looks like there's already a PR: https://github.com/imageworks/OpenColorIO/pull/551 and some discussion: https://github.com/imageworks/OpenColorIO/pull/518 on the subject of Werror. Not clear if/when they'll get round to doing something about it though.

For completeness, I opened an issue as well: https://github.com/imageworks/OpenColorIO/issues/559

Comment (Thra11) [reply pkgs/development/libraries/opencolorio/default.nix]: I've replaced the CXX_FLAGS expression which disabled errors on unused-function and deprecated-declarations with a postPatch that strips both occurences of -Werror.

**Comment by jtojnar** [comment]
@GrahamcOfBorg build opencolorio

**Pull request merged**